src/protocol-handlers/urls.js

```diff
@@ -0,0 +1,147 @@
+const toUri = require('multiaddr-to-uri')
+const { CID } = require('multiformats/cid')
+const fetch = require('node-fetch')
+
+const DEFAULT_GATEWAY = 'https://dweb.link'
+const LOCAL_HOSTNAMES = ['127.0.0.1', '[::1]', '0.0.0.0', '[::]']
+
+/**
+ * @typedef ParsedUrl
+ * @type {object}
+ * @property {string} protocol
+ * @property {string} hostname
+ * @property {string} path
+ */
+
+/**
+ * Parses an IPFS/IPNS/dWeb URL to be handled by IPFS Desktop.
+ *
+ * @param {string} url
+ * @returns {ParsedUrl}
+ */
+function parseUrl (url) {
+  let protocol = null
+  let hostname = null
+  let path = '/'
+
+  if (url.startsWith('ipfs://')) {
+    protocol = 'ipfs'
+    hostname = url.slice(7)
+  } else if (url.startsWith('ipns://')) {
+    protocol = 'ipns'
+    hostname = url.slice(7)
+  } else if (url.startsWith('dweb:/ipfs/')) {
+    protocol = 'ipfs'
+    hostname = url.slice(11)
+  } else if (url.startsWith('dweb:/ipns/')) {
+    protocol = 'ipns'
+    hostname = url.slice(11)
+  } else {
+    return null
+  }
+
+  if (hostname.includes('/')) {
+    const [first, ...rest] = hostname.split('/')
+    hostname = first
+    path = '/' + rest.join('/')
+  }
+
+  return { protocol, hostname, path }
+}
+
+async function getPublicGatewayUrl (ctx) {
+  if (!ctx.webui) {
+    // Best effort. If the Web UI window wasn't created yet, we just return the default
+    // gateway.
+    return DEFAULT_GATEWAY
+  }
+
+  return await ctx.webui
+    .webContents
+    .executeJavaScript('JSON.parse(localStorage.getItem("ipfsPublicGateway")) || "https://dweb.link"', true)
+}
+
+async function getPrivateGatewayUrl (ctx) {
+  const ipfsd = ctx.getIpfsd ? await ctx.getIpfsd(true) : null
+  if (!ipfsd || !ipfsd.api) {
+    return DEFAULT_GATEWAY
+  }
+
+  let gateway = await ipfsd.api.config.get('Addresses.Gateway')
+  if (Array.isArray(gateway)) {
+    if (gateway.length >= 1) {
+      gateway = gateway[0]
+    } else {
+      return DEFAULT_GATEWAY
+    }
+  }
+
+  return toUri(gateway)
+}
+
+const checkIfGatewayUrlIsAccessible = async (url) => {
+  try {
+    const { status } = await fetch(
+    `${url}/ipfs/bafkqae2xmvwgg33nmuqhi3zajfiemuzahiwss`
+    )
+    return status === 200
+  } catch (e) {
+    return false
+  }
+}
+
+// Separate test is necessary to see if subdomain mode is possible,
+// because some browser+OS combinations won't resolve them:
+// https://github.com/ipfs/go-ipfs/issues/7527
+const checkIfSubdomainGatewayUrlIsAccessible = async (url) => {
+  try {
+    url = new URL(url)
+    url.hostname = `bafkqae2xmvwgg33nmuqhi3zajfiemuzahiwss.ipfs.${url.hostname}`
```

Comment:
What is the hardcoded CID in `bafkqae2xmvwgg33nmuqhi3zajfiemuzahiwss.ipfs.${url.hostname}`?

Should that be saved in consts.js?

Reply:
Same as before: I did not write this code, but it seems that it is some safe CID (https://github.com/ipfs/ipfs-webui/commit/ebda7644dbf1a4cf0712b31a5fa3a68ccf3aa470) for "Welcome to IPFS :-)"

Reply:
This is a CID which has `Welcome to IPFS :-)` string inlined inside  `identity`  Multihash.
Main reason for using this exotic multihash type is to avoid hitting DHT/datastore – gateway can respond immediatelly after decoding CID.